src/object_store/src/object/s3.rs

```diff
@@ -228,16 +247,31 @@ impl StreamingUploader for S3StreamingUploader {
         fail_point!("s3_finish_streaming_upload_err", |_| Err(
             ObjectError::internal("s3 finish streaming upload error")
         ));
-        if self.join_handles.is_empty() && self.buf.is_empty() {
-            self.abort().await?;
-            return Err(ObjectError::internal("upload empty object"));
-        }
-        if let Err(e) = self.flush_and_complete().await {
+        // If the multipart upload has not been initiated, we can use `Put` instead to save the
+        // `CreateMultipartUpload` and `CompleteMultipartUpload` requests.
+        if self.upload_id.is_none() {
+            debug_assert!(self.join_handles.is_empty());
+            if self.buf.is_empty() {
+                debug_assert_eq!(self.not_uploaded_len, 0);
+                Err(ObjectError::internal("upload empty object"))
+            } else {
+                self.client
```

Comment:
I think we can move the `put_object` logic into `flush_and_complete` to make the `finish` look simpler, since falling back to `put_object` is a part of `flush_and_complete`.

Reply:
If `put_object` is moved into `flush_and_complete`, the `upload_id` needs to be checked twice. Besides, `flush_and_complete` was intended to complete the multipart upload process. Maybe we can rename `flush_and_complete` to `finish_stream_and_complete` to reflect that.

Reply:
The current code also checks whether `upload_id` is None for twice, once in `finish` and once in `upload_next_part` called by `flush_and_complete`.

Besides, `flush_and_complete` has the semantic of flushing anything left in the buffer to S3, without indicating how the data is flushed. But it's fine to rename it to `finish_multipart_and_complete` to indicate that this is only for finishing multipart.